### ISSUE TITLE
[Bug]: CDAT migration: AttributeError: 'CoreParameter' object has no attribute 'test_start_yr'

### ISSUE THREAD

**Issue by chengzhuzhang** [E3SM-Project/e3sm_diags#877]
### What happened?

When testing a model vs obs run when I tried to specify the start and end year for **reference data**, I ran into AttributeError for all parameter objects, example: AttributeError: 'CoreParameter' object has no attribute 'test_start_yr'

The same script works okay for `main` branch.

### Minimal Complete Verifiable Example (MVCE)

```python
import os
from e3sm_diags.parameter.core_parameter import CoreParameter
from e3sm_diags.run import runner

param = CoreParameter()

param.reference_data_path = '/global/cfs/cdirs/e3sm/diagnostics/observations/Atm/time-series'
param.test_data_path = '/global/cfs/cdirs/e3sm/chengzhu/eamxx/post/data/rgr'
param.test_name = 'eamxx_decadal'
param.seasons = ["ANN"]
#param.save_netcdf = True

param.ref_timeseries_input = True
# Years to slice the ref data, base this off the years in the filenames.
param.ref_start_yr = "1996"
param.ref_end_yr = "1996"

prefix = '/global/cfs/cdirs/e3sm/www/zhang40/tests/eamxx'
param.results_dir = os.path.join(prefix, 'eamxx_decadal_1996_1024_edv2')

runner.sets_to_run = ["lat_lon",
        "zonal_mean_xy",
        "zonal_mean_2d",
        "zonal_mean_2d_stratosphere",
        "polar",
        "cosp_histogram",
        "meridional_mean_2d",
        "annual_cycle_zonal_mean",]

runner.run_diags([param])
```

### Relevant log output

```python
(/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy) chengzhu@perlmutter:login35:~/eamxx_diags/run_script> python run_e3sm_diags_1996.py 
2024-10-25 14:41:09,034 [INFO]: e3sm_diags_driver.py(_save_env_yml:58) >> Saved environment yml file to: /global/cfs/cdirs/e3sm/www/zhang40/tests/eamxx/eamxx_decadal_1996_1024_edv3/prov/environment.yml
2024-10-25 14:41:13,375 [INFO]: e3sm_diags_driver.py(_save_parameter_files:69) >> Saved command used to: /global/cfs/cdirs/e3sm/www/zhang40/tests/eamxx/eamxx_decadal_1996_1024_edv3/prov/cmd_used.txt
2024-10-25 14:41:13,384 [INFO]: e3sm_diags_driver.py(_save_python_script:133) >> Saved Python script to: /global/cfs/cdirs/e3sm/www/zhang40/tests/eamxx/eamxx_decadal_1996_1024_edv3/prov/run_e3sm_diags_1996.py
2024-10-25 14:41:15,655 [ERROR]: core_parameter.py(_run_diag:343) >> Error in e3sm_diags.driver.lat_lon_driver
Traceback (most recent call last):
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/parameter/core_parameter.py", line 340, in _run_diag
    single_result = module.run_diag(self)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/lat_lon_driver.py", line 65, in run_diag
    test_ds = Dataset(parameter, data_type="test")
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/utils/dataset_xr.py", line 113, in __init__
    self.start_yr = self.parameter.test_start_yr  # type: ignore
AttributeError: 'CoreParameter' object has no attribute 'test_start_yr'
2024-10-25 14:41:15,696 [ERROR]: core_parameter.py(_run_diag:343) >> Error in e3sm_diags.driver.lat_lon_driver
Traceback (most recent call last):
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/parameter/core_parameter.py", line 340, in _run_diag
    single_result = module.run_diag(self)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/lat_lon_driver.py", line 65, in run_diag
    test_ds = Dataset(parameter, data_type="test")
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/utils/dataset_xr.py", line 113, in __init__
    self.start_yr = self.parameter.test_start_yr  # type: ignore
AttributeError: 'CoreParameter' object has no attribute 'test_start_yr'
```

### Environment

The latest cdat-migration-fy24 branch.

**Comment by tomvothecoder** [comment]
Thanks for reporting this. I will take a look today.

**Comment by tomvothecoder** [comment]
I think the issue is that on `cdat-migration-fy24`, the `self.start_yr` and `self.end_yr` are attributes that always expect valid values:

https://github.com/E3SM-Project/e3sm_diags/blob/cae04aa0126a2d178ae263678f25f3cc919e9c3a/e3sm_diags/driver/utils/dataset_xr.py#L99-L114

While on `main`, they are not class attributes in `dataset.py` but rather only returned in the function below. It also uses `getattr()` which returns `None` as a default value.
https://github.com/E3SM-Project/e3sm_diags/blob/4cda0a807e496a97392c14d92429bd614c536e4d/e3sm_diags/driver/utils/dataset.py#L236-L256


### Possible Workaround
Use `getattr()` for `CoreParameter` attributes in the code block below:
https://github.com/E3SM-Project/e3sm_diags/blob/cae04aa0126a2d178ae263678f25f3cc919e9c3a/e3sm_diags/driver/utils/dataset_xr.py#L99-L114

**Comment by tomvothecoder** [comment]
> It also uses getattr() which returns None as a default value.

Nevermind, `getattr()` does not automatically return `None` as a default value (has to be explicitly set).

I think on `main`, `self.parameter.test_start_yr` and `self.parameter.test_end_yr` are being set earlier before `self.start_yr` and `self.end_yr` are set so it works. More debugging needed.

**Comment by tomvothecoder** [comment]
Findings:
- On `main`: 
  - The `test_data` object is considered `is_climo` because it is `self.test` and `test_timeseries_input=False` in the script (logic below):
  - https://github.com/E3SM-Project/e3sm_diags/blob/4cda0a807e496a97392c14d92429bd614c536e4d/e3sm_diags/driver/utils/dataset.py#L188-L195
- On `cdat-migration-fy24`:
  - The `test_ds` object is considered `is_time_series` because it is `self.test` and `self.parameter.ref_timeseries_input` is set:
  - https://github.com/E3SM-Project/e3sm_diags/blob/cae04aa0126a2d178ae263678f25f3cc919e9c3a/e3sm_diags/driver/utils/dataset_xr.py#L133-L138

Question: Should `test_data`/`test_ds` be considered `is_time_series`? And if yes, should `test_timeseries_input=True` in the run script? I noticed that `main` sets the the ref climo variable to the test climo variable (model-only run).

**Comment by chengzhuzhang** [comment]
@tomvothecoder thank you for troubleshooting. I think the logic on `main` is correct. 
The `test/ref_timeseries_input` are default to `False`. 

if `test/ref_timeseries_input` is set to `True`, then the corresponding test/ref is considered is_time_series. 

if a dataset is considered to `is_time_series` or not is independent of the other dataset.

**Comment by tomvothecoder** [comment]
Closed by #881
